### PR TITLE
Loop count for all trivial form loops. Mem2Reg only when necessary.

### PR DESCRIPTION
**Pull request by adam-yang**
Using Scalar Evolution to figure out trivial loop trip count, but it needs Mem2reg to have been done, so we can't just run promote allocas as part of the unroll pass.

This change adds a pass to conditionally run m2r on functions that have loops that need to be unrolled when Od option is used.

### CODE REVIEW

File: tools/clang/test/CodeGenHLSL/batch/declarations/precise/matrix.hlsl [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %dxc -E main -T vs_6_0 %s | FileCheck %s | XFail GitHub #2080
+// RUN: %dxc -E main -T vs_6_0 %s | FileCheck %s
```
Comment (MrTrillian) [inline tools/clang/test/CodeGenHLSL/batch/declarations/precise/matrix.hlsl]: How could your change fix this???

Comment (adam-yang) [reply tools/clang/test/CodeGenHLSL/batch/declarations/precise/matrix.hlsl]: DxilConditionalMem2Reg only promotes allocas if it isn't marked precise.

Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/batch/declarations/precise/matrix.hlsl]: I don't see that in your code? These things should be orthogonal.

Comment (adam-yang) [reply tools/clang/test/CodeGenHLSL/batch/declarations/precise/matrix.hlsl]: It's on line 661 of DxilLoopUnroll.cpp in the custom Mem2Reg routine. This was used exclusively for unroll situations but now it's for general uses.

**Pull request merged**